src/tensors/tensor.jl

```diff
@@ -771,20 +589,21 @@ function Base.convert(::Type{TensorMap}, t::AbstractTensorMap)
     return copy!(TensorMap{scalartype(t)}(undef, space(t)), t)
 end
 
-function Base.convert(TT::Type{<:TensorMap{T,S,N₁,N₂}},
-                      t::AbstractTensorMap{<:Any,S,N₁,N₂}) where {T,S,N₁,N₂}
+function Base.convert(TT::Type{TensorMap{T,S,N₁,N₂,A}},
+                      t::AbstractTensorMap{<:Any,S,N₁,N₂}) where {T,S,N₁,N₂,A}
     if typeof(t) === TT
         return t
     else
-        data = Dict{sectortype(TT),storagetype(TT)}(c => convert(storagetype(TT), b)
-                                                    for (c, b) in blocks(t))
-        return TensorMap(data, codomain(t), domain(t))
+        data = convert(A, t.data)
+        return TensorMap(data, space(t))
     end
 end
 
 function Base.promote_rule(::Type{<:TT₁},
                            ::Type{<:TT₂}) where {S,N₁,N₂,
                                                  TT₁<:TensorMap{<:Any,S,N₁,N₂},
                                                  TT₂<:TensorMap{<:Any,S,N₁,N₂}}
-    return tensormaptype(S, N₁, N₂, promote_type(storagetype(TT₁), storagetype(TT₂)))
+    T = promote_type(scalartype(TT₁), scalartype(TT₂))
+    A = promote_type(storagetype(TT₁), storagetype(TT₂))
```

Comment:
I am not sure if this can ever pose problems, but we could ensure compatibility between T and A by:
```suggestion
    A = promote_type(storagetype(TT₁), storagetype(TT₂))
    T = scalartype(A)
```

Reply:
Neither of the two are probably a good idea:
```julia-repl
julia> promote_type(Vector{Float64}, Vector{ComplexF32})
Vector (alias for Array{T, 1} where T)
```

Reply:
https://discourse.julialang.org/t/type-promototion-of-vectors/30227/6
https://github.com/JuliaLang/julia/blob/e52a46c5c192bfe16853ae0b63ac33b280fba063/base/range.jl#L1282

It seems like this is not really something that the Base promotion system is designed to resolve. From
what I can find, the current behaviour is to only promote if at least one of the eltypes would not change, which is probably not what we want.

I would maybe suggest inserting our own `promote_storagetype`, which defaults to `Vector{promote_scalartype}` and we can then further specialise however we want? It seems like this is something that would be resolved with the new `Memory` types too, as we can then just hardcode that and only have to deal with the different `addrspace` implementations for GPU/CPU...

Reply:
I see no logic in the response of StefanKarpinski to Matt's question. Maybe we can just use `VectorInterface.promote_add(A1, A2)`?

Reply:
That indeed seems to work, I think that should be fine. It does rely on type inference, but I guess we can see if it causes any issues and still change it if we have to?